chore: Add smoke test

marikaner opened the pull request:
Add smoke tests using the sample code.

tomfrenken commented on .github/workflows/smoke-test.yml:
```diff
@@ -0,0 +1,40 @@
+          node-version: ${{ vars.DEFAULT_NODE_VERSION }}
+          cache: 'pnpm'
+      - run: pnpm i --frozen-lockfile
+      # - name: Create .env file
```
I suppose this can be removed, since the binding will always be in the VCAP variables?

tomfrenken commented on tests/smoke-test/.cfignore:
```diff
@@ -0,0 +1,4 @@
+coverage/
+jest.config.mjs
+tsconfig.json
+.env
```
+new line

marikaner replied:
```suggestion
.env

```

tomfrenken commented on tests/smoke-test/package.json:
```diff
@@ -0,0 +1,28 @@
+    "cf-login": "dotenv -- bash -c 'cf login -a \"$CF_API_URL\" -u \"$CF_USER\" -p \"$CF_PASSWORD\" -o \"$CF_ORG\" -s \"$CF_SPACE\"'",
+    "test": "NODE_OPTIONS=--experimental-vm-modules jest",
+    "lint": "prettier . --config ../../.prettierrc --ignore-path ../../.prettierignore -c",
+    "lint:fix": "prettier . --config ../../.prettierrc --ignore-path ../../.prettierignore -w --log-level error"
```
I belive eslint is missing from lint:fix

tomfrenken replied:
I guess we also missed that in the e2e-tests directory, unless it was intentional

MatKuhr replied:
I think for e2e tests I just copied the lint command from somewhere else. Is this not the right command? 😄

tomfrenken replied:
It's partially correct, the follow command chains with eslint, here is the command from one of our packages:
```
"eslint . --fix && prettier . --config ../../.prettierrc --ignore-path ../../.prettierignore -w --log-level error
```

tomfrenken commented on tests/smoke-test/tsconfig.json:
```diff
@@ -0,0 +1,11 @@
+{
+  "extends": "../../tsconfig.json",
+  "compilerOptions": {
+    "rootDir": "./deployment",
```
there is no `./deployment` directory, I suppose this is just a placeholder since there are no ts files to be transpiled?

marikaner replied:
it is being created on compile. not sure if this is the best practice, but the code is copied from the sample code, therefore it should not be duplicated.

tomfrenken replied:
I see, I think I would prefer it if we use the existing files instead and adapt the package.json scripts accordingly

tomfrenken reviewed (dismissed):
Some clean-up here and there, overall LGTM

MatKuhr commented on package.json:
```diff
@@ -11,14 +11,18 @@
     "ai-core": "pnpm -F=@sap-ai-sdk/ai-core",
     "gen-ai-hub": "pnpm -F=@sap-ai-sdk/gen-ai-hub",
-    "core": "pnpm -F=@sap-ai-sdk/core"
+    "core": "pnpm -F=@sap-ai-sdk/core",
```
can't find this option in the docs, what does `-F` do?

deekshas8 replied:
It filters packages based on the provided selector. https://pnpm.io/filtering

deekshas8 commented on tests/smoke-test/jest.config.mjs:
```diff
@@ -0,0 +1,5 @@
+import config from '../../jest.config.mjs';
+export default {
+  ...config,
```
the config here should be same as the e2e-tests. The global setup/teardown should not be called. Same for the tsconfig.

marikaner replied:
What do you mean by "Same for the tsconfig."

deekshas8 replied:
I meant the additional testconfig.test.json that we added for unit testing:

The transform option should be:
```
transform: {
    '^.+\\.tsx?$': [
      'ts-jest',
      {
        useESM: true,
      },
    ],
  },
```

and not (from the common jest.config.mjs)
```
transform: {
    '^.+\\.tsx?$': [
      'ts-jest',
      {
        useESM: true,
        tsconfig: path.resolve(__dirname, 'tsconfig.test.json'),
      },
    ],
  }
  ```

deekshas8 commented on tests/smoke-test/test/smoke.test.ts:
```diff
@@ -0,0 +1,16 @@
+const smokeTestRoute =
+  'https://smoke-test-app.cfapps.eu12-001.hana.ondemand.com';
+describe('Smoke Test', () => {
+  it('aicore', async () => {
```
```suggestion
  it('gets a 200 response from aicore service endpoint', async () => {
```

marikaner replied:
I changed it, but used different test. Your suggestion actually showed me that I am not testing what I think I am testing ;)

MatKuhr commented on tests/smoke-tests/.cfignore:
```diff
@@ -0,0 +1,6 @@
+node_modules/
+jest.config.mjs
+tsconfig.json
+.env
```
What about the dist folder? Does that have any impact?

marikaner replied:
I want it to be pushed explicitly, I only copy the dist folder at the moment

MatKuhr commented on tests/smoke-tests/README.md:
```diff
@@ -0,0 +1,38 @@
+
+## Running the app locally
+
+
```
Leftover?

marikaner replied:
I didn't finish 🤦‍♀️

MatKuhr commented on tests/smoke-tests/package.json:
```diff
@@ -0,0 +1,27 @@
+    "start": "node dist/server.js",
+    "add-canary": "pnpm add @sap-ai-sdk/ai-core@canary --save-prod=false --save-dev=false && pnpm add @sap-ai-sdk/gen-ai-hub@canary --save-prod=false --save-dev=false",
+    "create-deployment": "pnpm add-canary && rm -rf dist && cp ../../sample-code/dist dist",
+    "cf-login": "dotenv -- bash -c 'cf login -a \"$CF_API_URL\" -u \"$CF_USER\" -p \"$CF_PASSWORD\" -o \"$CF_ORG\" -s \"$CF_SPACE\"'",
```
Why dotenv?

marikaner replied:
It sets the env variables from the .env, so you can reference them in the command after this. Of course you can login manually, but I found this much more convenient

tomfrenken commented on .pnpmfile.cjs:
```diff
@@ -0,0 +1,19 @@
+module.exports = {
+  hooks: {
+    afterAllResolved: async lockfile => {
+      // Object.keys(lockfile.packages)
```
Uhm, leftover?

marikaner replied:
Nope, I want to keep this for the follow up, because this code will most likely be needed for it.

tomfrenken commented on tests/smoke-tests/README.md:
```diff
@@ -0,0 +1,48 @@
+
+The purpose of the smoke tests is to find obvious issues in the e2e process of using the SAP Cloud SDK for AI libs.
+
+The sample code app is running on Cloud Foundry and uses the current canary version of the SAP Cloud SDK for AI.
```
```suggestion
The sample application runs on Cloud Foundry and uses the current canary version of the SAP Cloud SDK for AI.
```

marikaner replied:
(I believe `is running` is better to express that it is running all the time)

tomfrenken commented on tests/smoke-tests/README.md:
```diff
@@ -0,0 +1,48 @@
+
+## Deploying a new version of the sample code app
+
+The sample code app is deployed to SAP BTP nightly.
```
```suggestion
The sample application is deployed to SAP BTP nightly.
```

tomfrenken commented on tests/smoke-tests/README.md:
```diff
@@ -0,0 +1,48 @@
+## Deploying a new version of the sample code app
+
+The sample code app is deployed to SAP BTP nightly.
+To deploy a new version of the sample code app manually, make sure to provide your own `.env` file in the `test/smoke-tests` subdirectory with the following environment variables:
```
```suggestion
To deploy a new version of the sample application manually, make sure to provide your own `.env` file in the `test/smoke-tests` subdirectory with the following environment variables:
```

tomfrenken reviewed (changes requested):
Overall looks pretty straightforward, however, there seem to be some leftovers, LGTM post-cleanup.

marikaner commented:
> Also, the pipeline seems to fail because cloud foundry installs node 18, I suppose you need to adjust the manifest.yaml to select node 20.

@tomfrenken  I wonder where you got this idea? To me it seems that I deleted a script during clean up. I added it back, should work now.

deekshas8 commented on sample-code/src/server.ts:
```diff
@@ -35,8 +36,26 @@ app.get('/embedding', async (req, res) => {
+
+app.get('/ai-core/get-deployments', async (req, res) => {
+  try {
+    res.send(await chatCompletion());
```
Is this correct? shouldn't we call getDeployments function?

marikaner replied:
oopsie, good catch

marikaner commented on tests/smoke-tests/package.json:
```diff
@@ -0,0 +1,27 @@
+    "@types/express": "^4.17.21",
+    "dotenv": "^16.4.5",
+    "dotenv-cli": "^7.4.2"
+  }
```
```suggestion
  },
  "engines": {
    "node": "^20"
  }
```

marikaner commented on tests/smoke-tests/package.json:
```diff
@@ -0,0 +1,30 @@
+    "@types/express": "^4.17.21",
+    "dotenv": "^16.4.5",
+    "dotenv-cli": "^7.4.2"
+  },
+  "engines": {
+    "node": "^20"
```
```suggestion
```

marikaner commented on tests/smoke-tests/package.json:
```diff
@@ -0,0 +1,30 @@
+    "test": "NODE_OPTIONS=--experimental-vm-modules jest",
+    "lint": "eslint . && prettier . --config ../../.prettierrc --ignore-path ../../.prettierignore -c",
+    "lint:fix": "eslint . --fix && prettier . --config ../../.prettierrc --ignore-path ../../.prettierignore -w --log-level error"
+  },
```
```suggestion
    },
  "engines": {
    "node": "^20"
    },
```

tomfrenken reviewed (approved):
LGTM
